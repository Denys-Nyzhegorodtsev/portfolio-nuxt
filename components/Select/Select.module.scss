.customSelect {
  position: relative;
  font-family: Arial, sans-serif;
  border-style: solid;
  border-width: 1px;
  border-color: transparent;
  border-radius: 8px 8px 0 0;
  cursor: pointer;
  padding: 10px;

  &.open {
    border-color: var(--color-gray-100);
    .options {
      top: 100%;
      opacity: 1;
      pointer-events: all;
    }
  }

  &.close {
    border-color: transparent;
    .options {
      opacity: 0;
      top: -100%;
      transition: all ease-in-out 0.3s;
      pointer-events: none;
    }
  }
}

.selectedOption {
  font-size: 16px;
  background-color: var(--color-gray-50);
  cursor: pointer;
}

.options {
  position: absolute;
  left: -1px;
  border: 1px solid var(--color-gray-100);
  border-radius: 0 0 8px 8px;
  border-top: none;
  max-height: 150px;
  overflow-y: auto;
  transition: all ease-in-out 0.3s;
  z-index: 1000;
}

.option {
  padding: 10px;
  cursor: pointer;
  transition: background-color 0.3s;
  border-bottom: 1px solid var(--color-gray-100);

  &:last-child {
    border: none;
  }

  &:hover {
    background-color: var(--color-gray-default);
  }
}
