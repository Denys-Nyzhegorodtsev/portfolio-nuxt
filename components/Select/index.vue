<template>
  <div
    :class="[
      'custom-select',
      $style.customSelect,
      isOpen ? $style.open : $style.close,
    ]"
  >
    <div :class="$style.selectedOption" @click="toggleOptions">
      {{ selectedLabel }}
    </div>
    <div :class="$style.options">
      <div
        v-for="option in options"
        :key="option.value"
        :class="$style.option"
        @click="selectOption(option)"
      >
        {{ option.label }}
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
  import {
    ref,
    onMounted,
    onBeforeUnmount,
    defineProps,
    defineEmits,
    watch,
  } from 'vue';

  interface Option {
    label: string;
    value: string;
  }

  // Приймаємо пропс та еміт для v-model
  const props = defineProps<{ modelValue: string; options: Option[] }>();
  const emit = defineEmits(['update:modelValue']);

  const isOpen = ref(false);
  const selectedLabel = ref('Select an option');
  const options = props.options;

  // Оновлюємо `selectedLabel` коли змінюється `modelValue`
  watch(
    () => props.modelValue,
    (newValue) => {
      const selectedOption = options.find(
        (option) => option.value === newValue
      );
      if (selectedOption) {
        selectedLabel.value = selectedOption.label;
      }
    },
    { immediate: true }
  );

  const handleClickOutside = (event: MouseEvent) => {
    const element = event.target;
    if (element instanceof HTMLElement && !element.closest('.custom-select')) {
      isOpen.value = false;
    }
  };

  const toggleOptions = () => {
    isOpen.value = !isOpen.value;
  };

  const selectOption = (option: Option) => {
    selectedLabel.value = option.label;
    emit('update:modelValue', option.value); // Оновлюємо значення v-model
    isOpen.value = false;
  };

  onMounted(() => {
    document.addEventListener('click', handleClickOutside);
  });

  onBeforeUnmount(() => {
    document.removeEventListener('click', handleClickOutside);
  });
</script>

<style module type="scss" src="./Select.module.scss"></style>
