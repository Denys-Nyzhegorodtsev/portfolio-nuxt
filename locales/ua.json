{
  "hello": "Добрий день, {name}!",
  "language": "Мова"
}
