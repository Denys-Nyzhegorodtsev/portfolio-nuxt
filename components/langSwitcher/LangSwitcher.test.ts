import { describe, it, expect } from 'vitest';
import { mount, VueWrapper } from '@vue/test-utils';
import LangSwitcher from './index.vue';
import { createRouter, createMemoryHistory } from 'vue-router';
import { createI18n } from 'vue-i18n';
import { nextTick } from 'vue';
import Select from '../Select/index.vue';
import en from '../../locales/en.json';
import ua from '../../locales/ua.json';

const router = createRouter({
  history: createMemoryHistory(),
  routes: [{ path: '/:locale', component: LangSwitcher }],
});

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: { en, ua },
});

describe('LangSwitcher Component', () => {
  const wrapper = mount(LangSwitcher, {
    global: {
      plugins: [router, i18n],
    },
  });

  it('renders correctly with snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('updates locale and selectedLocale when route param changes', async () => {
    const router = createRouter({
      history: createMemoryHistory(),
      routes: [{ path: '/:locale', component: LangSwitcher }],
    });

    const i18n = createI18n({
      legacy: false,
      locale: 'en',
      messages: { en, ua },
    });

    const wrapper = mount(LangSwitcher, {
      global: {
        plugins: [router, i18n],
      },
    });

    // Перевіряємо початкове значення
    expect(i18n.global.locale.value).toBe('en');
    expect((wrapper.vm as any).selectedLocale).toBe('en');

    // Імітуємо зміну параметра `locale` в маршруті на `ua`
    await router.push('/ua');
    await router.isReady();
    await nextTick();

    // Перевірка оновлених значень
    expect(i18n.global.locale.value).toBe('ua');
    expect((wrapper.vm as any).selectedLocale).toBe('ua');
  });
});
