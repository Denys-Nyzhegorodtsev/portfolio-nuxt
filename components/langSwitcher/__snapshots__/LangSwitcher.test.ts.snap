// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LangSwitcher Component > renders correctly with snapshot 1`] = `
"<div class="custom-select _customSelect_8795fe _close_8795fe">
  <div class="_selectedOption_8795fe">en</div>
  <div class="_options_8795fe">
    <div class="_option_8795fe">en</div>
    <div class="_option_8795fe">ua</div>
  </div>
</div>"
`;
